### PR TITLE
[sca] KMAC uJSON handler

### PR DESCRIPTION
**Pull request by nasahlpa**
This PR translates the existig simpleserial KMAC SCA tests to the uJSON communication framework.

### CODE REVIEW

File: sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+ */
+uint8_t key_fixed[kKeyLength];
+
+/**
+ * Fixed-key indicator.
+ *
+ * Used in the 'b' (batch capture) command for indicating whether to use fixed
+ * or random key.
+ */
+static bool run_fixed = false;
```
Comment (m-temp) [inline sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c:83]: For AES this is set to true. Is it important that the first run is the same amongst all scripts, or do we not care about this?
If we do care: should this be factored out in a separate PR (in both repos, here and in the ot-sca)?

**Review by m-temp — APPROVED** [review]
I'm not an expert with respect to the cryptotest framework and the uJSON part.
From the KMAC perspective it looks good.

File: sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+   * Number of cycles (at `kClockFreqCpuHz`) that Ibex should sleep to minimize
+   * noise during SHA3 operations. Caution: This number should be chosen to
+   * provide enough time. Otherwise, Ibex might wake up while SHA3 is still busy
+   * and disturb the capture. Currently, we use a start trigger delay of 320
+   * clock cycles and the scope captures 160 clock cycles at kClockFreqCpuHz
+   * (3200 samples). On the scope side, an offset of 395 clock cycles (7900
+   * samples) is used to ignore 1) the start trigger delay for the PROCESS
+   * command, and 2) the absorb phase for the fixed prefix.
+   */
+  kIbexSha3SleepCycles = 1180,
```
Comment (vogelpi) [inline sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c:44]: Also here, I would remove the samples numbers (see AES and SHA3).

File: sw/device/tests/crypto/cryptotest/json/kmac_sca_commands.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_TESTS_CRYPTO_CRYPTOTEST_JSON_KMAC_SCA_COMMANDS_H_
+#define OPENTITAN_SW_DEVICE_TESTS_CRYPTO_CRYPTOTEST_JSON_KMAC_SCA_COMMANDS_H_
+#include "sw/device/lib/ujson/ujson_derive.h"
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define KMACSCA_CMD_MAX_BATCH_DIGEST_BYTES 128
```
Comment (vogelpi) [inline sw/device/tests/crypto/cryptotest/json/kmac_sca_commands.h:12]: Should this be 64 instead or is there a use case when we feedback 128 Bytes?

**Review by vogelpi — APPROVED** [review]
This looks good with with two minor comments. Thanks for working on this @nasahlpa and @m-temp !

One thing that came to my mind when reviewing this: On the FPGA, we made a couple of modifications to the RTL beyond the trigger delay to ease pre-silicon SCA. This includes:
- writing the MSG FIFO before the prefix/key is being hashed
- providing two commands back to back

This won't be possible with silicon. We should think of how to adapt the `kmac_sca.c/h` to support both targets or think about removing the pre-silicon mode completely. I'll file an issue.

File: sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c (hunk 2 above)
Comment (nasahlpa) [reply sw/device/tests/crypto/cryptotest/firmware/kmac_sca.c]: Done

**Comment by nasahlpa** [comment]
Thanks for your comments @vogelpi & @m-temp!

Regarding the pre-silicon mode, I think we should address this in a separate PR where we adapt all the uJSON based cipher capture programs accordingly. Thanks for creating an issue in this context!
